flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecSink.java

```diff
@@ -647,15 +649,51 @@ private InternalTypeInfo<RowData> getInputTypeInfo() {
     }
 
     private int[] getPrimaryKeyIndices(RowType sinkRowType, ResolvedSchema schema) {
-        return schema.getPrimaryKey()
-                .map(k -> k.getColumns().stream().mapToInt(sinkRowType::getFieldIndex).toArray())
-                .orElse(new int[0]);
+        if (schema.getPrimaryKey().isPresent()) {
+            UniqueConstraint uniqueConstraint = schema.getPrimaryKey().get();
+            int[] primaryKeyIndices = new int[uniqueConstraint.getColumns().size()];
+            // sinkRowType may not contain full primary keys in case of row-level update or delete.
+            // in such case, return an empty array
+            for (int i = 0; i < uniqueConstraint.getColumns().size(); i++) {
+                int fieldIndex = sinkRowType.getFieldIndex(uniqueConstraint.getColumns().get(i));
+                if (fieldIndex == -1) {
+                    return new int[0];
+                }
```

Comment:
In here, we do validate all avaliable pk columns.
The logic is try to find all the pk columns from the `ResolvedSchema`, but if for one pk column, we find we can't find from the `ResolvedSchema`,  that means the pk columns  miss some columns in which case we will consider it as no pk columns.
Note: this case only happens in update statement, the required columns returned by connector don't contain all pk columns.

Reply:
I mean why we don't return all available positions here?

```
  uniqueConstraint.getColumns().stream()
                    .mapToInt(sinkRowType::getFieldIndex)
                    .filter(i -> i == -1)
                    .toArray();
```

I find the indcies is used as shuffle key, in the update statement we don't need shuffle?

Reply:
We use primary keys as shuffle keys, but they still are primary keys which maybe used for other purposes in the `CommonExecSink`.  If not all primary keys are available, we should consider it as no primary keys which is the semantic of primary keys.
Btw, I have moved such logic to `BatchExecSink` to make every clear since only update statement which is only supported in batch  needs to consider it.